packages/protocol/contracts/L1/libs/LibUtils.sol

```diff
@@ -6,141 +6,51 @@
 
 pragma solidity ^0.8.18;
 
+import {ChainData} from "../../common/IXchainSync.sol";
+import {LibMath} from "../../libs/LibMath.sol";
+import {LibTokenomics} from "./LibTokenomics.sol";
 import {
     SafeCastUpgradeable
 } from "@openzeppelin/contracts-upgradeable/utils/math/SafeCastUpgradeable.sol";
-
-import {LibMath} from "../../libs/LibMath.sol";
 import {TaikoData} from "../TaikoData.sol";
 
 library LibUtils {
     using LibMath for uint256;
 
-    bytes32 public constant BLOCK_DEADEND_HASH = bytes32(uint256(1));
-
     error L1_BLOCK_NUMBER();
 
-    struct StateVariables {
-        uint256 feeBase;
-        uint64 genesisHeight;
-        uint64 genesisTimestamp;
-        uint64 nextBlockId;
-        uint64 lastProposedAt;
-        uint64 avgBlockTime;
-        uint64 latestVerifiedHeight;
-        uint64 latestVerifiedId;
-        uint64 avgProofTime;
-    }
-
-    function getProposedBlock(
-        TaikoData.State storage state,
-        uint256 maxNumBlocks,
-        uint256 id
-    ) internal view returns (TaikoData.ProposedBlock storage) {
-        return state.proposedBlocks[id % maxNumBlocks];
-    }
-
-    function getL2BlockHash(
+    function getL2ChainData(
         TaikoData.State storage state,
         uint256 number,
         uint256 blockHashHistory
-    ) internal view returns (bytes32) {
-        if (
-            number + blockHashHistory <= state.latestVerifiedHeight ||
-            number > state.latestVerifiedHeight
+    ) internal view returns (ChainData storage) {
+        uint256 _number = number;
+        if (_number == 0) {
+            _number = state.lastBlockId;
+        } else if (
+            _number + blockHashHistory <= state.lastBlockId ||
+            _number > state.lastBlockId
         ) revert L1_BLOCK_NUMBER();
 
-        return state.l2Hashes[number % blockHashHistory];
+        return state.l2ChainDatas[_number % blockHashHistory];
```

Comment:
I think we should add the check. I added a TODO for myself to do that. I think with the current storage layout, it's not easy. I'm going to organize all info related to a block to one struct and get this TODO done.

Reply:
I think if you move everything to the block struct, we'll have the chaindata for all blocks for "free" because the prover has to store them anyway and so we can easily read the data from those prover stored values without any more copying (so should actually be cheaper).

Reply:
The previous ChainData are only written for verified blocks, while the large ring buffer is for proposed/unverified blocks. In the worse case, all buffer are taken by unverified blocks, then all ChainData are gone for verified blocks. In the current implementation, I make it explicit that one ring buffer is called proposedBlock, and the other one is called verifiedBlocks.